packages/fuels-accounts/src/provider.rs

```diff
@@ -233,9 +237,55 @@ impl Provider {
         Ok(self.client.submit(&tx.into()).await?)
     }
 
+    #[cfg(feature = "coin-cache")]
+    async fn find_in_cache<'a>(
+        &self,
+        coin_ids: impl IntoIterator<Item = (&'a (Bech32Address, AssetId), &'a Vec<CoinTypeId>)>,
+    ) -> Option<((Bech32Address, AssetId), CoinTypeId)> {
+        let mut locked_cache = self.cache.lock().await;
+
+        for (key, ids) in coin_ids {
+            let items = locked_cache.get_active(key);
+
+            if items.is_empty() {
+                continue;
+            }
+
+            for id in ids {
+                if items.contains(id) {
+                    return Some((key.clone(), id.clone()));
+                }
+            }
+        }
+
+        None
+    }
+
+    #[cfg(feature = "coin-cache")]
+    async fn check_inputs_already_in_cache<'a>(
+        &self,
+        coin_ids: impl IntoIterator<Item = (&'a (Bech32Address, AssetId), &'a Vec<CoinTypeId>)>,
+    ) -> Result<()> {
+        use fuels_core::types::errors::{transaction, Error};
+
+        if let Some(((addr, asset_id), coin_type_id)) = self.find_in_cache(coin_ids).await {
+            let msg = match coin_type_id {
+                CoinTypeId::UtxoId(utxo_id) => format!("coin with utxo_id: `{utxo_id:x}`"),
+                CoinTypeId::Nonce(nonce) => format!("message with nonce: `{nonce}`"),
+            };
+            Err(Error::Transaction(transaction::Reason::Validation(
+                format!("{msg} already in cache. Wallet address: `{addr}`, asset id: `{asset_id}`"),
```

Comment:
I feel like just stating that the coin is already cached doesn't convey enough info to the user as to what is going on, especially if they are not fully aware of the internals of the SDK and the caveats of a UTXO based chain.

Saying something like:

"a transaction with this input was submitted recently, attempting to spend it again will result in an error".

should give a better idea of what the issue is.